web/packages/teleport/src/Roles/RoleEditor/StandardEditor/validation.ts

```diff
@@ -272,6 +302,19 @@ const noWildcardOptions = (message: string) => (options: Option[]) => () => {
   return { valid, message: valid ? '' : message };
 };
 
+const validateServerAccess = (
+  a: ServerAccess
+): ServerAccessValidationResult => {
+  const result = runRules(a, serverAccessValidationRules);
+  if (a.labels.length === 0 && a.logins.length === 0) {
+    result.valid = false;
+    result.message = 'At least one label or login required';
+    result.fields.labels.valid = false;
+    result.fields.logins.valid = false;
+  }
```

Comment:
Just wondering, shouldn't this be a part of `serverAccessValidationRules`? 
Technically, we are validating the value here.

Reply:
Not exactly: each rule is run only on the single data point that it validates (i.e. every field separately). Our validation library doesn't allow passing additional context. So the validation that we're doing here (and in other sections) is the validation of the _section_ itself. The reason I'm marking particular fields as invalid is for visual representation purposes (literally: making the inputs red). Notice that I'm only setting the `valid` field, not the messages, since the message is global, on the section level.